Datastore/src/Operation.php

```diff
@@ -871,4 +961,275 @@ private function sortEntities(array $entities, array $keys)
+                        : null,
+                    'namespace_id' => isset($key['partitionId']['namespaceId'])
+                        ? $key['partitionId']['namespaceId']
+                        : null,
+                    'database_id' => isset($key['partitionId']['databaseId'])
+                        ? $key['partitionId']['databaseId']
+                        : null,
+                ]);
+
+                $local['partition_id'] = new PartitionId($p);
+            }
+
+            $local['path'] = [];
+            if (isset($key['path'])) {
+                foreach ($key['path'] as $element) {
+                    $local['path'][] = new PathElement($element);
+                }
+            }
+
+            $out[] = $local;
+        }
+
+        return $out;
+    }
+
+    /**
+     * Parse query into API compatible values
+     *
+     * @param array $query
+     * @return array
+     */
+    private function parseQuery(array $query)
+    {
+        if (isset($query['order'])) {
+            foreach ($query['order'] as &$order) {
+                $order['direction'] = $order['direction'] === 'ASCENDING'
+                    ? Direction::ASCENDING
+                    : Direction::DESCENDING;
+            }
+        }
+
+        if (isset($query['filter'])) {
+            $query['filter'] = $this->convertFilterProps($query['filter']);
+        }
+
+        if (isset($query['limit']) && !is_array($query['limit'])) {
+            $query['limit'] = [
+                'value' => $query['limit']
+            ];
+        }
+
+        return $query;
+    }
+
+    /**
+     * Parse query into API compatible values
+     *
+     * @param array $gqlQuery
+     * @return array
+     */
+    private function parseGqlQuery(array $gqlQuery)
+    {
+        if (isset($gqlQuery['namedBindings'])) {
+            foreach ($gqlQuery['namedBindings'] as $name => &$binding) {
+                if (!isset($binding['value'])) {
+                    continue;
+                }
+
+                $binding = $this->prepareQueryBinding($binding);
+            }
+        }
+
+        if (isset($gqlQuery['positionalBindings'])) {
+            foreach ($gqlQuery['positionalBindings'] as &$binding) {
+                if (!isset($binding['value'])) {
+                    continue;
+                }
+
+                $binding = $this->prepareQueryBinding($binding);
+            }
+        }
+
+        return $gqlQuery;
+    }
+
+    /**
+     * Convert Query filters to an API-compatible value.
+     *
+     * @param array $filter The input filter data
+     * @return array
+     */
+    private function convertFilterProps(array $filter)
+    {
+        if (isset($filter['propertyFilter'])) {
+            $operator = $filter['propertyFilter']['op'];
+
+            try {
+                if (is_int($operator)) {
+                    // verify that the operator, given as enum value, exists.
+                    PropertyFilterOperator::name($operator);
```

Comment:
do we really need to do this? The API will validate this.

Reply:
Good catch Brent! This has been a redundant library level exception throwing as the backend already does that. We should be able to remove this.